### ISSUE TITLE
nvmf/rdma: Intermittent connect issues between kernel initiator and SPDK nvmf target

### ISSUE THREAD

**Issue by mhae** [spdk/spdk#2493]
# Bug report

Connecting to a large number of subsystems (~24) via the kernel initiator to the SPDK nvmf target fails intermittently:

When I connect to the remote target, I sometimes get this timeout error.
```
root@localhost:~# nvme connect-all -t rdma -a 192.168.1.1 -s 4420
Failed to write to /dev/nvme-fabrics: Connection timed out
```

The corresponding kernel log shows this:
```
[7684390.374707] nvme nvme1: new ctrl: NQN "nqn.2014-08.org.nvmexpress.discovery", addr 192.168.1.1:4420
[7684390.393557] nvme nvme1: Removing ctrl: NQN "nqn.2014-08.org.nvmexpress.discovery"
[7684390.543504] nvme nvme1: creating 16 I/O queues.
[7684391.114846] nvme nvme1: new ctrl: NQN "nqn.2019-20.x.y.z:0x11223344556677.0.bay00", addr 192.168.1.1:4420
[7684391.145837] nvme nvme2: creating 16 I/O queues.
[7684391.726776] nvme nvme2: new ctrl: NQN "nqn.2019-20.x.y.z:0x11223344556677.0.bay01", addr 192.168.1.1:4420
[7684391.757861] nvme nvme3: creating 16 I/O queues.
[7684392.328063] nvme nvme3: new ctrl: NQN "nqn.2019-20.x.y.z:0x11223344556677.0.bay02", addr 192.168.1.1:4420
[7684392.359152] nvme nvme4: creating 16 I/O queues.
[7684392.929321] nvme nvme4: new ctrl: NQN "nqn.2019-20.x.y.z:0x11223344556677.0.bay03", addr 192.168.1.1:4420
[7684392.960346] nvme nvme5: creating 16 I/O queues.
[7684393.530385] nvme nvme5: new ctrl: NQN "nqn.2019-20.x.y.z:0x11223344556677.0.bay04", addr 192.168.1.1:4420
[7684393.561376] nvme nvme6: creating 16 I/O queues.
[7684394.033608] nvme nvme6: new ctrl: NQN "nqn.2019-20.x.y.z:0x11223344556677.0.bay05", addr 192.168.1.1:4420
[7684394.064552] nvme nvme7: creating 16 I/O queues.
[7684394.634412] nvme nvme7: new ctrl: NQN "nqn.2019-20.x.y.z:0x11223344556677.0.bay06", addr 192.168.1.1:4420
[7684394.665357] nvme nvme8: creating 16 I/O queues.
[7684395.236514] nvme nvme8: new ctrl: NQN "nqn.2019-20.x.y.z:0x11223344556677.0.bay07", addr 192.168.1.1:4420
[7684395.267431] nvme nvme9: creating 16 I/O queues.
[7684395.838295] nvme nvme9: new ctrl: NQN "nqn.2019-20.x.y.z:0x11223344556677.0.bay08", addr 192.168.1.1:4420
[7684395.869244] nvme nvme10: creating 16 I/O queues.
[7684396.388194] nvme nvme10: new ctrl: NQN "nqn.2019-20.x.y.z:0x11223344556677.0.bay09", addr 192.168.1.1:4420
[7684396.430362] nvme nvme11: creating 16 I/O queues.
[7684396.989949] nvme nvme11: new ctrl: NQN "nqn.2019-20.x.y.z:0x11223344556677.0.bay10", addr 192.168.1.1:4420
[7684397.030565] nvme nvme12: creating 16 I/O queues.
[7684397.631249] nvme nvme12: new ctrl: NQN "nqn.2019-20.x.y.z:0x11223344556677.0.bay11", addr 192.168.1.1:4420
[7684397.662322] nvme nvme13: creating 16 I/O queues.
[7684398.242544] nvme nvme13: new ctrl: NQN "nqn.2019-20.x.y.z:0x11223344556677.0.bay12", addr 192.168.1.1:4420
[7684398.273441] nvme nvme14: creating 16 I/O queues.
[7684398.784614] nvme nvme14: new ctrl: NQN "nqn.2019-20.x.y.z:0x11223344556677.0.bay13", addr 192.168.1.1:4420
[7684398.835237] nvme nvme15: creating 16 I/O queues.
[7684399.435797] nvme nvme15: new ctrl: NQN "nqn.2019-20.x.y.z:0x11223344556677.0.bay14", addr 192.168.1.1:4420
[7684399.466817] nvme nvme16: creating 16 I/O queues.
[7684400.036593] nvme nvme16: new ctrl: NQN "nqn.2019-20.x.y.z:0x11223344556677.0.bay15", addr 192.168.1.1:4420
[7684400.077462] nvme nvme17: creating 16 I/O queues.
[7684403.274579] nvme nvme17: rdma connection establishment failed (-110)
```

I've also seen this error when I used -i 64 on the nvme connect-all command line:
```
[7685645.268934] nvme nvme3: Connect rejected: status 28 (consumer defined) nvme status 6 (resource not found).
[7685645.288488] nvme nvme3: rdma connection establishment failed (-104)
```
When I get the above error, nvmf_tgt shows one of these messages:
```
[2022-05-02 16:05:17.343240] rdma.c:1237:nvmf_rdma_event_accept: *ERROR*: Error 101 on spdk_rdma_qp_accept
[2022-05-02 16:05:17.343305] nvmf.c: 856:_nvmf_poll_group_add: *ERROR*: Unable to add the qpair to a poll group.
```
or
```
[2022-05-02 16:07:02.348681] rdma.c:1237:nvmf_rdma_event_accept: *ERROR*: Error 22 on spdk_rdma_qp_accept
[2022-05-02 16:07:02.348736] nvmf.c: 856:_nvmf_poll_group_add: *ERROR*: Unable to add the qpair to a poll group.
```

I see these messages as well right before the error occurs:
```
[2022-05-03 08:58:47.373001] rdma.c:3363:nvmf_process_ib_event: *NOTICE*: Async event: GID table change
[2022-05-03 08:58:47.373053] rdma.c:3363:nvmf_process_ib_event: *NOTICE*: Async event: GID table change
[2022-05-03 08:58:47.373069] rdma.c:3363:nvmf_process_ib_event: *NOTICE*: Async event: GID table change
[2022-05-03 08:58:47.373087] rdma.c:3363:nvmf_process_ib_event: *NOTICE*: Async event: GID table change
[2022-05-03 08:58:47.373102] rdma.c:3363:nvmf_process_ib_event: *NOTICE*: Async event: GID table change
[2022-05-03 08:58:47.373117] rdma.c:3363:nvmf_process_ib_event: *NOTICE*: Async event: GID table change
[2022-05-03 08:58:47.373134] rdma.c:3363:nvmf_process_ib_event: *NOTICE*: Async event: GID table change
[2022-05-03 08:58:47.373149] rdma.c:3363:nvmf_process_ib_event: *NOTICE*: Async event: GID table change
```

## Expected Behavior

The initiator should be able to connect to all subsystems without indicating an error.

## Current Behavior

I've seen this connection issue with the SPDK initiator as well. Right now, when the kernel initiator gets an error, it won't continue to connect to the rest of the subsystems. Nor will it retry.
The same problem happens periodically when I connect to each subsystem one at a time.
The problem doesn't always happen but still quite frequently (1 out of 3 attempts).

## Steps to Reproduce

```
nvmf_tgt -m 0xf
```

Script to set up:
```
#!/bin/bash
/tmp/scripts/rpc.py bdev_nvme_set_hotplug -e
/tmp/scripts/rpc.py nvmf_create_transport -t RDMA -u 131072 -f
sleep 5

for (( i=0; i<24; i++ ))
do
  bay=$(printf "%02d" $i)
  ns=$(printf "HotInNvme%dn1" $i)
  /tmp/scripts/rpc.py nvmf_subsystem_create nqn.2019-20.x.y.ze:0x11223344556677.0.bay$bay -a
  /tmp/scripts/rpc.py nvmf_subsystem_add_listener nqn.2019-20.x.y.z:0x11223344556677.0.bay$bay -t rdma -a 192.168.1.1 -s 4420
  /tmp/scripts/rpc.py nvmf_subsystem_add_ns nqn.2019-20.x.y.z:0x11223344556677.0.bay$bay $ns
done
```
Note, I'm using the insert_strip option (-f) but I don't think it matters. It also fails when I remove it or when I use the SPDK initiator that supports PRACT=0.

## Context (Environment including OS version, SPDK version, etc.)

Debian 10.10 (kernel version 4.19.171). The 2 nodes are direct connected with a CX5.
I used SPDK main as of 5/2/22 but it happens with older versions as well.

**Comment by jimharris** [comment]
[Bug scrub] Alexey: do you see anything in dmesg on target side? 
Michael: no

Alexey: usually we see these GID table change messages when some change is made on the physical port
Michael: I'm not making any changes to physical port while this test is running

Michael: could this be related to DHCP?
Alexey: it's possible, IP address change could cause this kind of message

Alexey: I'll take a look at this issue

Jim: have you tried testing this against kernel target?
Michael: I have not tried that yet

**Comment by mhae** [comment]
I think I found the issue... when I set up the IP address for the device via nmcli/nmtui, I did not change the "method" setting (ip4.method). It stayed at "automatic". After changing it to "manual", I no longer see the "Async event: GID table change" messages and the connects work fine.
Now as a test, I changed the method back to "automatic" and everything continues to work fine (no "Async event: GID table change" messages).  So I'm not really sure what was going on and why the networking stack is behaving that way.
Good news is that it seems to be a networking stack config issue, rather than RDMA/SPDK issue.
